File: C/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int calc_readabil(string text);

int main(void)
{
    //Get text from user and calculate the Coleman-Liau index
    string text = get_string("Text: ");
    int grade_level = calc_readabil(text);

    // Check to see if Coleman-Liau index is less than 1 or greater than 16, and print the index
    if (grade_level < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade_level >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade_level);
    }
}

int calc_readabil(string text)
{
    // Count the number of letters, words, and sentences in the text
    int letters = 0, words = 0, sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        int num = (int) toupper(text[i]);
        if (num >= 65 && num <= 90)
        {
            letters++;
        }
        else if (num == 32)
        {
            words++;
        }
        else if (num == 46 || num == 33 || num == 63)
        {
            sentences++;
            if (i == n - 1)
            {
                words++;
            }
        }
    }
    // Calculate the Coleman-Liau index
    float cl_float = 0.0588 * ((float) letters / (float) words * 100.0) - 0.296 * ((float) sentences / (float) words * 100.0) - 15.8;
    int cl_int = (int) round(cl_float);

    return cl_int;
}

File: Python/readability.py
from cs50 import get_string
import re


def main():

    # Get the name of the user
    text = get_string("Text: ")

    # Move through every character in the text, adding 1 to the counters for letters, words, and sentences when appropriate
    letters = 0
    words = 0
    sentences = 0
    for c in text:
        if re.match(r"[A-Z]", c.upper()):
            letters += 1
        elif re.match(r" ", c):
            words += 1
        elif re.match(r"[\!\?\.]", c):
            sentences += 1

    # Check to see if text ends in a space; if not, increase the number of words by 1
    if re.match(r" ", text[len(text) - 1]) == None:
        words += 1

    # Calculate the index
    index = 0.0588 * (letters / words * 100) - 0.296 * (sentences / words * 100) - 15.8

    # Check to see if the index < 1 or index >= 16, and print appropriate response
    if index <= 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        # Add 0.5 to index to ensure computer will round correctly
        print(f"Grade {int(index + 0.5)}")


if __name__ == "__main__":
    main()